File: path_sum_3.rb
# def path_sum(root, sum, running_sum = 0)
#     # res = dfs(root, sum, 0)
#     # res == nil ? 0 : res
#     return 0 unless root
#     new_sum = root.val + running_sum
#     to_add_to_total = new_sum == sum ? 1 : 0
#     return [
#         path_sum(root.left, sum, new_sum),
#         path_sum(root.right, sum, new_sum),
#         path_sum(root.left, sum, running_sum),
#         path_sum(root.right, sum, running_sum),
#         to_add_to_total
#     ].sum
# end

# def dfs(root, sum, running_sum)
#     return 0 unless root
#     new_sum = root.val + running_sum
#     to_add_to_total = new_sum == sum ? 1 : 0
#     return [
#         dfs(root.left, sum, new_sum),
#         dfs(root.right, sum, new_sum),
#         dfs(root.left, sum, running_sum),
#         dfs(root.right, sum, running_sum),
#         to_add_to_total
#     ].sum
# end

def path_sum(root, sum)
    @count = 0
    dfs(root, sum)
    @count
end

def dfs(root, target)
    return unless root
    @count += 1 if root.val + subtotal == @target
    dfs(root.left, 0) # ()
    dfs(root.left, root.val)
    dfs(root.left, subtotal + root.val) # ()
    dfs(root.right, 0) # ()
    dfs(root.right, root.val)
    dfs(root.right, subtotal + root.val) # ()
end

def test(root, target)

end

File: pathSum3.py
# You are given a binary tree in which each node contains an integer value.

# Find the number of paths that sum to a given value.

# The path does not need to start or end at the root or a leaf, but it must go downwards (traveling only from parent nodes to child nodes).

# The tree has no more than 1,000 nodes and the values are in the range -1,000,000 to 1,000,000.

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# root = [10,5,-3,3,2,null,11,3,-2,null,1], sum = 8

#       10
#      /  \
#     5   -3
#    / \    \
#   3   2   11
#  / \   \
# 3  -2   1

# Return 3. The paths that sum to 8 are:

# 1.  5 -> 3
# 2.  5 -> 2 -> 1
# 3. -3 -> 11

class Solution:
    def __init__(self):
        self.count = 0 # ++

    def pathSum(self, root: TreeNode, sum: int) -> int: # 10
        self.dfs(root, sum)
        return self.count

    def dfs(self, root, target): # 11
        if not root: return 0
        left1 = root.val + self.pathSum(root.left, sum) # 10 +  // -3 + // 11 + 0
        left2 = self.pathSum(root.left, sum) # // 11
        right1 = root.val + self.pathSum(root.right, sum) # 18 + // 8 //
        right2 = self.pathSum(root.right, sum) # // 11
        
        for s in [left1, left2, right1, right2]:
            if s == target: self.count += 1

        return root.val
        # return self.count


#       10
#      /  \
#     5   -3
#    / \    \
#   3   2   11
#  / \   \
# 3  -2   1

        # return a sum at the end
        
        # test if current node.val + pathSum() == 8
        # AND test if pathSum() == 8



    # def someMethod(self, someParam=None)
        








